authz/rbac_translator.go

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package authz exposes methods to manage authorization within gRPC.
+package authz
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	v3rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	v3routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	v3matcherpb "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+)
+
+type header struct {
+	Key    string   `json:"key"`
+	Values []string `json:"values"`
+}
+
+type peer struct {
+	Principals []string `json:"principals"`
+}
+
+type request struct {
+	Paths   []string `json:"paths"`
+	Headers []header `json:"headers"`
+}
+
+type rule struct {
+	Name    string  `json:"name"`
+	Source  peer    `json:"source"`
+	Request request `json:"request"`
+}
+
+// Represents the SDK authorization policy provided by user.
+type authorizationPolicy struct {
+	Name       string `json:"name"`
+	DenyRules  []rule `json:"deny_rules"`
+	AllowRules []rule `json:"allow_rules"`
+}
+
+func principalOr(principals []*v3rbacpb.Principal) *v3rbacpb.Principal {
+	return &v3rbacpb.Principal{
```

Comment:
Should all 4 of these functions special case `if len(input)==1 { return input[0] }`?

Reply:
I feel we should avoid the special case. Here the parsing is based on the SDK policy semantics. So the expected behavior is (consider request field), that it will be an AND of paths and headers. Paths will be ORed together. Headers has its own semantics. Now if the slice has only one url path, then we skip the OR part, and directly include URL PATH to AND, functionally its correct (and also simplified, reduced nesting), but I would still prefer having strict structure. Especially as the policy grows and adds more fields.